### ISSUE TITLE
Global variable marked `eslint(no-undef)` in the Vue script section

### ISSUE THREAD

**Issue by JensDll** [vuejs/eslint-plugin-vue#2089]
**Checklist**
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**
- **ESLint version:**  v8.34.0
- **eslint-plugin-vue version:** v9.9.0
- **Node version:** v18.13.0
- **Operating System:** Windows

**Please show your full configuration:**
```yaml
root: true
env:
  browser: true
  es2022: true
  node: true
parser: vue-eslint-parser
parserOptions:
  parser: '@typescript-eslint/parser'
  sourceType: 'module'
plugins:
  - '@typescript-eslint'
extends:
  - eslint:recommended
  - plugin:@typescript-eslint/recommended
  - plugin:vue/vue3-recommended
ignorePatterns:
  - node_modules
  - dist
```

**What did you do?**
```ts
// In a TypeScript declaration file
declare const MY_GLOBAL: string
```
```vue
<!-- In a Vue file -->
<script setup lang="ts">
console.log(MY_GLOBAL)
</script>

<template>
  <h1>eslint(no-undef)</h1>
</template>
```

**What did you expect to happen?**
No undeclared variable warning. 

**What actually happened?**
```text
App.vue
  2:13  error  'MY_GLOBAL' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)
```

Please note I'm opening this here because the same code is valid in scripts outside of `.vue` files.

**Repository to reproduce this issue**
https://github.com/JensDll/eslint-plugin-vue-no-undef

**Comment by FloEdelmann** [comment]
You also have to tell ESLint about global variables via the `globals` section in the config or a special comment. it should then work in js/ts files as well as in Vue files.

**Comment by JensDll** [comment]
Ok, thanks, this is fine. I wondered, however, why the Vue script is lint differently than other TypeScript files. Initially, I saw this error (in Vue files) coming from TypeScript-declared definitions like [`PublicKeyCredential`](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential), which went away after configuring [`@typescript-eslint/parser`](https://typescript-eslint.io/architecture/parser) to include my `tsconfig`:

```yaml
parserOptions:
  parser: '@typescript-eslint/parser'
  sourceType: module
  project: ./tsconfig.json
  extraFileExtensions:
    - .vue
```
Please feel free to close this if you think this to be expected.

**Comment by FloEdelmann** [comment]
It might be because you have rules enabled that require type information? See https://typescript-eslint.io/linting/typed-linting

In any case, please open a new specific issue with a reproduction showing that problem.
